### ISSUE TITLE
3.1.3: vsetenv PMIX_SERVER_TMPDIR failed

### ISSUE THREAD

**Issue by benmenadue** [open-mpi/ompi#6068]
### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)

OpenMPI 3.1.3 + external PMIx 3.0.2

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

release tar-balls

### Please describe the system on which you are running

* Operating system/version: CentOS 6.10
* Computer hardware: various
* Network type: IB

-----------------------------

## Details of the problem

OpenMPI library is configured against an external PMIx build. Launching an MPI program produces the message
```
vsetenv PMIX_SERVER_TMPDIR failed
```
several times from each MPI process on calling `MPI_Init`.

A bit of `strace`-ing suggests the message is being generated by ibverbs -- all of the messages are generated follow this trace (or very similar, e.g. an extra `brk` call or three along the way):
```
[pid 20935] open("/dev/infiniband/uverbs0", O_RDWR) = 15
[pid 20935] write(15, "\0\0\0\0\4\0\6\0\200\3452W\377\177\0\0", 16) = 16
[pid 20935] mmap(NULL, 4096, PROT_WRITE, MAP_SHARED, 15, 0) = 0x2aebf1cea000
[pid 20935] mmap(NULL, 4096, PROT_WRITE, MAP_SHARED, 15, 0x1000) = 0x2aebf1ceb000
[pid 20935] write(15, "D\0\0\0\1\0=\0\220\3372W\377\177\0\0\0\0\0\0\0\0\0\0\16\20\0\0\0\0\0\0", 32) = 32
[pid 20935] ioctl(15, MGSL_IOCGPARAMS or MMTIMER_GETRES or MTIOCTOP or SNDCTL_MIDI_MPUMODE, 0x7fff5732e5dc) = 0
[pid 20935] mmap(NULL, 3928, PROT_READ, MAP_SHARED, 15, 0x3000) = 0x2aebf1cec000
[pid 20935] write(2, "vsetenv PMIX_SERVER_TMPDIR failed\n", 34vsetenv PMIX_SERVER_TMPDIR failed
) = 34
```
I suspect it's due to there being 2 copies of that environment variable (!) in the environment somehow:
```
> mpirun -np 1 strings /proc/self/environ | grep PMIX_SERVER_TMPDIR
PMIX_SERVER_TMPDIR=/jobfs/local/2185069.r-man2/ompi.r41.5800/pid.17479/0/0
PMIX_SERVER_TMPDIR=/jobfs/local/2185069.r-man2/ompi.r41.5800/pid.17479
```

**Comment by rhc54** [comment]
I'm not seeing that on master running on Mac OSX:
```shell
$ mpirun -n 1 env | grep TMPDIR
TMPDIR=/Users/rhc/tmp
PMIX_SERVER_TMPDIR=/Users/rhc/tmp/ompi.Ralphs-iMac-2.1000/pid.98053
PMIX_SYSTEM_TMPDIR=/Users/rhc/tmp
```
I would suspect a potential bug somewhere in OMPI 3.1.3 or else something weird about your system. I'll give it a try on a CentOS VM.

**Comment by rhc54** [comment]
Nope, works fine for master on CentOS 7 as well:
```shell
$ mpirun -n 1 env | grep TMPDIR
PMIX_SERVER_TMPDIR=/tmp/ompi.localhost.1000/pid.124127
PMIX_SYSTEM_TMPDIR=/tmp
```
Will test OMPI 3.1 next

**Comment by rhc54** [comment]
I'm a tad suspicious about the setup, but I can try to replicate it. Technically, OMPI 3.1.x doesn't support PMIx v3 though you can compile against it. Those envars, however, are set by ORTE and not PMIx, which makes the whole thing look fishy.

**Comment by benmenadue** [comment]
Yes, when I went looking I noticed that there was only an `ext2x` component in 3.1.3. But since it let me build it, I just thought I'd report it. If it's not a supported configuration, that's fine - easy enough to use PMIx v2 instead.

**Comment by benmenadue** [comment]
Found it, I think. There's this inside `orted` in the v3.1 branch: https://github.com/open-mpi/ompi/blob/17040631627b760efe3da95b4fc75eaf1d1aca4c/orte/orted/pmix/pmix_server.c#L251-L252

It was taken out of master by fa18ba395d72b7d584a65e69788c0af9d850cf4c.

**Comment by rhc54** [comment]
No, that line of code is fine - the ```true``` indicates that it should replace any existing value with the new one. Indeed, if you look down inside that code, all it does is call ```putenv``` with the new value. It was removed in master because we passed it in the attributes for server_init instead.

Besides, that code gets executed no matter which PMIx you built against, so you wouldn't expect to see a difference when switching from v2.x to v3.x. I searched the v3.x code and cannot find any place where we set that envar, so something really fishy is going on here.
